### PR TITLE
feat(x/swingset): include swing-store in genesis

### LINKED ISSUE #6527: tools to export/import swing-store to/from genesis block
## What is the Problem Being Solved?

A subset of the state-sync problem is to support emergency chain upgrades in which an old validator (which has halted) performs a "genesis block export", basically copying the contents the IAVL tree (only the most recent blockHeight/version) into a big JSON file. People can then edit the file manually (e.g. removing validators which are known to be wedged), then start a new chain from this modified genesis data. The new chain has a different name, and starts at a few blocks higher than the old one (to guard against accidental double-signs by validators who didn't stop at quite the right time). The cosmos operations for this are just `agd export` and `agd import` (or something similar).

Our chain keeps a lot of state outside the IAVL tree, so `agd export` won't automatically get enough state to launch a new validator.

## Description of the Design

The idea would be to add a swing-store API to dump the entire state of the swing-store out as key-value pairs. Basically we'd iterate through the `kvStore` and emit `kv-${key}: ${value}` for each entry, then walk the `streamStore` (transcript entries) and emit `stream-${vatID}-${transcriptNum}: ${transcriptEntry}` for each, then walk the `snapStore` (XS heap snapshots) and emit base64-encoded compressed snapshot files as strings.

This iteration must be in-consensus, which means:
* keys must be sorted in some canonical manner
* the `snapStore` might have extra snapshots that aren't being used by any vats (which will be deleted soon, but there are race conditions), so maybe the tool should be aware of the `kvStore` entries that indicate which snapshots are used by which vats, and only emit the within-consensus used ones
* the `kvStore` also contains `local.*` keys which are defined to be outside consensus, these should be omitted

Then we'd have a second API which can take these key-value pairs and initialize a new swing-store from them.

Then we modify `agd export` and `import` to use these APIs when building/consuming the genesis block.

This is like what we need for state-sync, but easier, because:
* export only happens while the node is halted, so we aren't trying to ignore ongoing mutations
* we can take as long as we need to generate the data; there's no block waiting for its hash
* this is only performed in an emergency, so even less frequently than daily-ish state-sync generation
* everybody agrees upon the same blockHeight to perform the export

## Security Considerations

Genesis export is a manual consensus operation: everybody can perform their own export and compare hashes, but there's no automatic staking/evidence/slashing taking place. On the plus side, any validators who launch from a different genesis block than the majority will fall out of consensus immediately.

## Test Plan

Unit tests in `packages/swing-store`, probably some unit tests in `cosmic-swingset` if we can come up with a good scheme.

cc @arirubinstein @JimLarson @michaelfig @FUDCo @mhofman

### PR DESCRIPTION
**Pull request by mhofman**
closes: #6527

## Description

This PR adds swing-store artifacts to the genesis export, and the ability to start from a genesis that includes a full swing-store export.

The artifacts are exported and imported from a directory alongside the `genesis.json` file, instead of being inlined like other state.

While the full swing-store state is now included in the genesis state, it doesn't mean it's editable in any kind of capacity. Most swingset state is likely replicated in multiple places inside the swing-store, like transcripts and xsnap heap snapshots.

### Security Considerations

As with any genesis start, the validators are in charge of verifying the validity of the genesis file against an expected hash, this PR does not change that.

They do not need to perform manual validation of the artifacts since those are checked for consistency by the swing-store import process, using metadata stored in the genesis file.

### Scaling Considerations

The swing-store artifacts are pretty large binary blobs, and as such are not included in the protobuf (genesis state) that ends up JSON serialized. However their size is still significant, and I believe a genesis export of mainnet would result in a total size of about 10GB today. Validators may want to exchange the genesis and artifacts compressed as that should compress down to about 800MB as seen with state-sync snapshots.

### Documentation Considerations

A new `--export-dir` option is required for `agd export` to specify a path to a directory where the genesis export process will create the `genesis.json` file and a `swing-store` directory containing the swing-store export artifacts. On start from genesis, this `swing-store` directory must be placed in the `config` folder of the node's home (alongside where the `genesis.json` usual reside), at least until InitGenesis has completed.

We do not expect this mechanism to be used for normal operations, just as a backstop in case of chain failure.

### Testing Considerations

Added a genesis test to the docker upgrade test framework.

### Upgrade Considerations

While this mechanism can be used for manual chain upgrades, it is fully transparent to any swingset operation.

### CODE REVIEW

**Comment by mhofman** [comment]
The size of the uncompressed artifacts is too large to be included in the genesis.json. We'll look into an alternative, likely a new option to `agd export` to indicate a directory path used to output everything.

**Comment by mhofman** [comment]
Best reviewed commit-by-commit

File: golang/cosmos/cmd/agd/main.go
Comment (michaelfig) [inline golang/cosmos/cmd/agd/main.go]: I dislike the use of a boolean to distinguish `agd start` from `agd export`, especially since then the names `OnStartHook` and `daemonize` are pretty misleading.  Can you instead just add `daemoncmd.OnExportHook` and leave the meaning of `daemoncmd.OnStartHook` alone?

This `main.go` could use this same closure for both:
```suggestion
	launchVM := func(logger log.Logger) {
		...
	}
	daemoncmd.OnExportHook = launchVM
	daemoncmd.OnStartHook = launchVM
```

File: golang/cosmos/cmd/libdaemon/main.go
Comment (michaelfig) [inline golang/cosmos/cmd/libdaemon/main.go]: As per the above in `agd/main.go`, please revert this change.

File: golang/cosmos/daemon/cmd/root.go
Comment (michaelfig) [inline golang/cosmos/daemon/cmd/root.go]:
```suggestion
var OnStartHook func(logger log.Logger)
var OnExportHook func(logger log.Logger)
```

File: golang/cosmos/daemon/cmd/root.go
Comment (michaelfig) [inline golang/cosmos/daemon/cmd/root.go]: Please revert.

File: golang/cosmos/daemon/cmd/root.go
Comment (michaelfig) [inline golang/cosmos/daemon/cmd/root.go]:
```suggestion
	if OnExportHook != nil {
		OnExportHook(logger)
	}
```

**Review by michaelfig — COMMENTED** [review]
A few comments.  I'll continue with review later.

File: golang/cosmos/daemon/cmd/root.go
Comment (michaelfig) [inline golang/cosmos/daemon/cmd/root.go:330]: What I've discerned is that `agd export --export-dir=foo` is a required flag, but `--swing-store-export-dir=foo/something` can be used to override where the swingstore is exported.  Is that true?  In either case, could you add this in comments near `FlagExportDir` and `FlagSwingStoreExportDir`?

File: golang/cosmos/daemon/cmd/root.go
Comment (michaelfig) [inline golang/cosmos/daemon/cmd/root.go]: I don't understand why this special case is necessary.  Isn't it more robust to have the same behaviour regardless of whether or not we're starting the VM?

```suggestion
	cmd.RunE = extendedRunE
```

**Review by michaelfig — APPROVED** [review]
LGTM (conditional on addressing comments)!  You probably should get a Golang wizard's review before landing, but I didn't see any obvious faux pas.

File: golang/cosmos/daemon/cmd/root.go
Comment (mhofman) [reply golang/cosmos/daemon/cmd/root.go:330]: FlagSwingStoreExportDir can only be used to set where to read the swing-store artifacts from for InitGenesis, not for export

File: golang/cosmos/daemon/cmd/root.go
Comment (mhofman) [reply golang/cosmos/daemon/cmd/root.go]: Without it we'd try to create the genesis file twice, and fail the second time around. I initially worked around that by truncating existing files

Comment (michaelfig) [reply golang/cosmos/daemon/cmd/root.go]: Ah, okay.  That's more subtle than I saw at first.  I suggest instead detecting whether `sender` has been set, and only running this function in that particular case, as I have commented above in this file.

Then, you can omit this condition based on `OnExportHook`, which is a separate concern from whether the current binary is executing a different one to obtain the controller (`sender == nil`).

File: golang/cosmos/daemon/cmd/root.go
Comment (michaelfig) [inline golang/cosmos/daemon/cmd/root.go]:
```suggestion
	// Only do the following modifications if we actually have a controller.
	// Otherwise, there will be an exec of the VM program, and it will
	// handle all of the export logic.
	if sender != nil {
		for _, command := range rootCmd.Commands() {
			if command.Name() != "export" {
				continue
			}

			extendCosmosExportCommand(command)
			break
		}
	}
```

File: golang/cosmos/daemon/cmd/root.go
Comment (mhofman) [reply golang/cosmos/daemon/cmd/root.go]: I did a variation on this as I think we should fail early if `--export-dir` is missing. PTAL

**Review by michaelfig — COMMENTED** [review]
Thanks, the new code looks good.

File: golang/cosmos/daemon/cmd/root.go
Comment (JimLarson) [inline golang/cosmos/daemon/cmd/root.go]: You can save a few lines by inverting the test and using the implicit `continue` at the end of the loop body.

```suggestion
		if command.Name() == "export" {
			extendCosmosExportCommand(command, sender != nil)
		         break
		}
```

File: golang/cosmos/daemon/cmd/root.go
Comment (JimLarson) [inline golang/cosmos/daemon/cmd/root.go:31]: I don't think `viper` provides enough functionality to justify its use. However, I see it's already used in cosmos-sdk, so consistency with existing code trumps my concerns.

File: golang/cosmos/daemon/cmd/root.go
Comment (JimLarson) [inline golang/cosmos/daemon/cmd/root.go:302]: Add note in the flag help string explaining difference from the swing store export dir flag, and in that one's help string explaining the difference from this one.

File: golang/cosmos/daemon/cmd/root.go
Comment (JimLarson) [inline golang/cosmos/daemon/cmd/root.go:301]: Document, briefly, what this function should do.

File: golang/cosmos/x/swingset/genesis.go
Comment (JimLarson) [inline golang/cosmos/x/swingset/genesis.go]:
```suggestion
		keeper.SwingStoreExportProvider{
			BlockHeight: snapshotHeight,
			GetExportDataReader: getExportDataReader,
			ReadNextArtifact: artifactProvider.ReadNextArtifact
		},
		keeper.SwingStoreRestoreOptions{
			ArtifactMode: keeper.SwingStoreArtifactModeReplay,
			ExportDataMode: keeper.SwingStoreExportDataModeAll
		},
```

**Review by JimLarson — APPROVED** [review]
Golang code LGTM with some minor requests and suggestions.

File: golang/cosmos/daemon/cmd/root.go
Comment (mhofman) [reply golang/cosmos/daemon/cmd/root.go:31]: Yeah this is just imported so I can cast to the `Viper` instance that cosmos-sdk creates and that's underlying the appOpts

File: golang/cosmos/daemon/cmd/root.go
Comment (mhofman) [reply golang/cosmos/daemon/cmd/root.go:302]: I didn't add a help string for `FlagSwingStoreExportDir` as from what I understand it could technically be set by config as well, not just by the command line. I have to admit I'm a little fuzzy with how the whole config thing is plumbed in cosmos-sdk. 

Actually I realize now that I didn't actually declare it as an official flag anywhere, I just abused the flag mechanism to plumb through an app option. Where do you recommend I declare the `FlagSwingStoreExportDir` flag?

File: golang/cosmos/cmd/agd/main.go
Comment (gibson042) [inline golang/cosmos/cmd/agd/main.go:33]: There's some log output in `launchVM` that could probably be improved for export.

File: golang/cosmos/daemon/cmd/root.go
Comment (gibson042) [reply golang/cosmos/daemon/cmd/root.go]: Hmm, `sender != nil` is a bit opaque. Maybe it should be assigned to a named variable with an explanatory comment?

File: golang/cosmos/daemon/cmd/root.go
Comment (gibson042) [inline golang/cosmos/daemon/cmd/root.go:245]: It's nice to see this named.

File: golang/cosmos/daemon/cmd/root.go
Comment (gibson042) [reply golang/cosmos/daemon/cmd/root.go:302]: I second the confusion, but `FlagSwingStoreExportDir` seems to have strange wiring. At minimum I would look for doc comments.

File: golang/cosmos/daemon/cmd/root.go
Comment (gibson042) [reply golang/cosmos/daemon/cmd/root.go:330]: Do they need to be different? It seems like the purpose is aligned.

File: golang/cosmos/daemon/cmd/root.go
Comment (gibson042) [inline golang/cosmos/daemon/cmd/root.go:333]: This will catch a preexisting genesis file, but is there also a potential issue if swingStoreExportPath was already created? I suspect not because it would be empty if this line succeeds, but it's worth confirming.

Comment (mhofman) [reply golang/cosmos/daemon/cmd/root.go:333]: If the `export-manifest.json` file exists in the swingStoreExportPath, the JS export process will fail. Directories existing are not a problem.

**Pull request merged**
